### DISCUSSION TITLE
Interactive Help in Pyodide Console

### DISCUSSION THREAD

**Discussion by tilman-schieber** [pyodide/pyodide#4123] [Q&A]
I am using the Pyodide Console for students to use interactively in an online course. 
It works great but when one enters the interactive help with `help()` there's a weird javascript popup:
![image](https://github.com/pyodide/pyodide/assets/986273/09f8fda1-b00a-451d-a62b-61058755309f)

I am running it on repl.it for testing (that's the URL shown in the dialogue) and the incomplete German text says : 
> "...of an in URL embedded page"

no matter what I enter, when I confirm the dialogue, the console informs me that I left the interactive help
~~~~~~~~
help> 
You are now leaving help and returning to the Python interpreter.
If you want to ask for help on a particular object directly from the
interpreter, you can type "help(object)".  Executing "help('string')"
has the same effect as typing a particular string at the help> prompt.`
~~~~~~~~

is there some way to get the interactive help to work or to disable it completely?
My only idea is to change the interpreter object passed to Jquery-terminal to ignore the `help()` command but that seems unnecessarily messy.

**Comment by ryanking13** [comment]
What happens when you call `help()` is that it tries to read user input from stdin. Pyodide calls [`prompt` Web API](https://developer.mozilla.org/en-US/docs/Web/API/Window/prompt) as a default handler for stdin, which is why it shows the JS prompt.

You can control the behavior of [stdin / stdout / stderr when loading Pyodide](https://pyodide.org/en/latest/usage/faq.html#how-can-i-control-the-behavior-of-stdin-stdout-stderr).

> is there some way to get the interactive help to work

It depends. Blocking the browser thread in the main thread is not available, so if you want to use a Python REPL-like interactive help, you will need to run Pyodide inside a web worker. This [presentation](https://www.youtube.com/watch?v=-SggWFS15Do&t=1750s) might help.

> or to disable it completely?

I think you can just override `help` with something else.

```js
let pyodide = await loadPyodide();
pyodide.globals.set("help", () => console.log("not supported"));
pyodide.runPython("help()");
```

**Reply by tilman-schieber** [reply]
Thank you for the great answer. 
I'll do some research into the first option but for now overriding help would suffice.
As I want the students to be able to use `help(str)` it would be great if it were possible to override only the interactive help.

in python I would do:
~~~~python
import builtins
def help(o=None):
    if o:
        builtins.help(o)
    else:
        print("not supported")
~~~~
I added this to the initial `runPython` command in `https://pyodide.org/en/stable/console.html` line 104
I thought that's the right place because other functions are defined there but unfortunately it's not in scope anymore once user input is interpreted.

That is confusing to me because I thought in the line

~~~~python
 pyconsole = PyodideConsole(__main__.__dict__)
~~~~
the current namespace is passed to the console, and the newly created help-function should be included. What am I missing?

**Reply by hoodmane** [reply]
I'd recommend money patching `builtins.help` then it'll be available everywhere.
